### ISSUE TITLE
Add native pointer overloads of methods that accept IntPtr

### ISSUE THREAD

**Issue by AArnott** [AArnott/pinvoke#76]
IntPtr parameters are harder to deal with than native pointer types (in C# anyway), and they express less in terms of type safety than pointers can. 

We should consider changing our policy and code to use pointers instead of IntPtr.

One possibility is that we maintain two overloads of applicable P/Invoke methods: one that takes IntPtr and one that takes pointers, so that folks can call without using `unsafe` code. I don't have much sympathy for any anxiety over writing unsafe code since it's actually simpler, and partial trust is irrelevant for consumers of this library anyway. So the only reason to keep IntPtr around is for languages that don't support pointers (e.g. Powershell).
- [x] AdvApi32
- [x] BCrypt
- [x] DbgHelp
- [x] Gdi32
- [x] Hid
- [x] ImageHlp
- [x] Kernel32
- [x] NCrypt
- [x] Psapi
- [x] SetupApi
- [x] User32

**Comment by AArnott** [comment]
CC: @vbfox

**Comment by vbfox** [comment]
Yes we need a list of all the pro/cons.

On the list of languages :
- :white_check_mark: C# (`unsafe` / `T* foo`)
- :white_check_mark: C++ (`T* foo`)
- :white_check_mark:  F# (`let foo : nativeptr<'T>`)
- :white_check_mark: Boo (`unsafe foo as T`) didn't test myself but they had a [unit test](https://github.com/bamboo/boo/blob/1f2f60a08ca69e95db34b30154985170f4723cad/tests/testcases/unsafe/unsafe-3.boo) about unsafe (Language is now considered dead)
- :large_blue_circle: VB.Net
- :large_blue_circle: PowerShell
- :8ball: Nemerle (?)
- :8ball: IronSheme (?)
- :8ball: IronPython (?)

:white_check_mark:: Can call pointer exposing method
:large_blue_circle:: Can call `IntPtr` overload even if pointer version exists
:interrobang: Doesn't support the `IntPtr` overload if the pointer version is present
:8ball:: Didn't find anythging / not searched

Also for the two overloads solutions it's pretty easy to codegen the `IntPtr` overloads if we write the pointer ones.

**Comment by AArnott** [comment]
Can you add a legend for these icons above?

**Comment by vbfox** [comment]
Ok I differentiated between VB that doesn't allow unsafe but can call the `IntPtr` version if also present with the same name and languages that support it natively.

Turn out that powershell is able to call an `IntPtr` version too if it exists (Just tested) with the same name so no problem here either.

Everything seem to point that the dual pointer/`IntPtr` is the good way to go, but we need to codegen it or it will be problematic pretty fast.

**Comment by vbfox** [comment]
Also if we decide to go the "dual" way it's not really needed for a 0.x release, as we will only add overloads that aren't compatible and don't risk being picked by previously existing code.
But It would be nice to have that in 1.0 along with some variant of #70.

**Comment by AArnott** [comment]
Thanks, @vbfox. I agree: let's go for dual mode. Until we have the code gen going, we should at _least_ offer IntPtr versions so we don't block out use by VB. 
Presumably when we start code gen that converts IntPtr to pointers, we can remove any manually created pointer overloads.

**Comment by AArnott** [comment]
I just found out that unsafe code isn't allowed in C# iterator methods. That means that pointers aren't allowed in them, making it less convenient to call these methods even in a language that would otherwise allow it.

**Comment by AArnott** [comment]
An open question on resolving some of these is what to do about struct fields that are pointers. Should they be IntPtr or a native pointer type? Can we perhaps use native pointers, but offer properties that expose them as IntPtrs for the VB folks?

**Comment by AArnott** [comment]
Progress on fixing kernel32 is blocked on AArnott/CodeGeneration.Roslyn#1

**Comment by jmelosegui** [comment]
Well, if we want to be consistent, struct fields should be native pointer types and the codegen should inject property wrappers of those fields.

So I am leaning toward native pointer type and properties that expose them as IntPtrs.

**Comment by AArnott** [comment]
Agreed.

**Comment by AArnott** [comment]
#130 tracks the struct fields now.
